dlt/load/load.py

```diff
@@ -414,72 +452,49 @@ def load_single_package(self, load_id: str, schema: Schema) -> None:
-                if len(remaining_jobs) == 0:
-                    # get package status
-                    package_info = self.load_storage.normalized_packages.get_load_package_info(
-                        load_id
-                    )
-                    # possibly raise on failed jobs
-                    if self.config.raise_on_failed_jobs:
-                        if package_info.jobs["failed_jobs"]:
-                            failed_job = package_info.jobs["failed_jobs"][0]
-                            raise LoadClientJobFailed(
-                                load_id,
-                                failed_job.job_file_info.job_id(),
-                                failed_job.failed_message,
-                            )
-                    # possibly raise on too many retries
-                    if self.config.raise_on_max_retries:
-                        for new_job in package_info.jobs["new_jobs"]:
-                            r_c = new_job.job_file_info.retry_count
-                            if r_c > 0 and r_c % self.config.raise_on_max_retries == 0:
-                                raise LoadClientJobRetry(
-                                    load_id,
-                                    new_job.job_file_info.job_id(),
-                                    r_c,
-                                    self.config.raise_on_max_retries,
-                                )
+                # we continously spool new jobs and complete finished ones
+                running_jobs, pending_exception = self.complete_jobs(load_id, running_jobs, schema)
+                # do not spool new jobs if there was a signal
+                if not signals.signal_received() and not pending_exception:
+                    running_jobs += self.spool_new_jobs(load_id, schema, len(running_jobs))
+                self.update_loadpackage_info(load_id)
+
+                if len(running_jobs) == 0:
+                    # if a pending exception was discovered during completion of jobs
+                    # we can raise it now
+                    if pending_exception:
+                        raise pending_exception
                     break
-                # process remaining jobs again
-                jobs = remaining_jobs
                 # this will raise on signal
-                sleep(1)
+                sleep(0.5)
             except LoadClientJobFailed:
                 # the package is completed and skipped
+                self.update_loadpackage_info(load_id)
                 self.complete_package(load_id, schema, True)
                 raise
 
+        # always update load package info
+        self.update_loadpackage_info(load_id)
+
+        # complete the package if no new or started jobs present after loop exit
+        if (
+            len(self.load_storage.list_new_jobs(load_id)) == 0
+            and len(self.load_storage.normalized_packages.list_started_jobs(load_id)) == 0
+        ):
```

Comment:
is this the correct "package completion" condition? I think so, but am not 100% sure.

Reply:
this was the previous condition:
```
        if file_count == 0:
            logger.info(f"No new jobs found in {load_id}")
            return 0, []
```
so checking just new jobs. I do not fully get why you do it here again? the loop above should exit only when all jobs are completed, no?